docs/dificil/small_to_large.md

```diff
@@ -1,7 +1,225 @@
+
+
+## Small to large folosind `__gnu_pbds`
+
+Această structură de date oferă aceleași funcții ca structura `set` din STL, însă mai oferă încă 2 funcții care ne pot fi de ajutor în anumite probleme:
+
+1. `find_by_order(x)`, care returnează un iterator către a x-a cheie ca ordine, în timp $O(log(n))$.
+2. `order_of_key(x)`, care returnează numărul de elemente **strict** mai mici ca x, tot în timp $O(log(n))$.
+
+Pentru a putea folosi această structură, va trebui să adăugăm la programul nostru următoarele linii:
+```c++
+#include <ext/pb_ds/assoc_container.hpp>
+#include <ext/pb_ds/tree_policy.hpp>
+using namespace __gnu_pbds;
+#define ordered_set tree<int, null_type, std::less_equal<int>, rb_tree_tag, tree_order_statistics_node_update>
+```
+În interiorul parantezelor ascuțite de după `tree`, primul câmp reprezintă tipul de date, în acest caz `int`, al patrulea reprezintă tipul de arbore folosit, în acest caz, `rb_tree`, care garantează complexitatea logaritmică a operațiilor de inserare și ștergere. Al treilea câmp reprezintă tipul de comparator folosit, unde `std::less_equal` permite existența a mai multor elemente egale (ca un `multiset`), iar `std::less` nu (ca un `set`).
+
+### Problemă: [E - Penultimul Cox -  Kilonova](https://kilonova.ro/problems/575)
+
+Această problemă se poate rezolva folosind o structură de tip `__gnu_pbds` care admite mai multe chei egale (folosește comparatorul `std::less_equal`). Este ușor să ne dăm seama că răspunsul pentru fiecare nod este chiar `s[nod].order_of_key(v[nod] + 1)`. Acel +1 se datorează faptului că ne trebuie numărul de elemente mai mici sau egale cu $v[nod]$, iar metoda `order_of_key` returnează numărul de elemente strict mai mici.
+
+O implementare a ideii de mai sus ar fi următoarea:
+
+```c++
+#include <iostream>
+#include <vector>
+
+#include <ext/pb_ds/assoc_container.hpp>
+#include <ext/pb_ds/tree_policy.hpp>
+using namespace __gnu_pbds;
+
+//avem elemente de tip int, care se repeta
+#define ordered_set tree<int, null_type, std::less_equal<int>, rb_tree_tag, tree_order_statistics_node_update>
+
+const int NMAX = 100005;
+std::vector<std::vector<int>> G(NMAX);
+ordered_set s[NMAX];
+int v[NMAX];
+int rez[NMAX];
+
+void dfs(int nod) {
+    for (auto x : G[nod]) {
+        dfs(x);
+        //la fel ca la set, operatia de swap ia tot O(1)
+        if (s[x].size() > s[nod].size())
+            s[x].swap(s[nod]);
+        for (auto x2 : s[x])
+            s[nod].insert(x2);
+    }
+    //numarul de elemente din set mai mici sau egale cu v[nod]
+    rez[nod] = s[nod].order_of_key(v[nod] + 1);
+}
+
+int main() {
+    int n, i, t;
+    std::ios_base::sync_with_stdio(false);
+    std::cin.tie(nullptr);
+    std::cin >> n;
+    for (i = 2; i <= n; i++) {
+        std::cin >> t;
+        G[t].push_back(i);
+    }
+    for (i = 1; i <= n; i++) {
+        std::cin >> v[i];
+        s[i].insert(v[i]);
+    }
+    dfs(1);
+    for (i = 1; i <= n; i++)
```

Comment:
Dacă ai putea să shiftezi să fie de la 0 la n, ai putea folosi mai consistent in for-urile tale.